### PR TITLE
refactor(datasource): migrate to class based datasource

### PR DESCRIPTION
**Pull request by JamieMagee**
A small experiment into what OOP/class based datasources might look like. Picked Cdnjs as it's the smallest & simplest.

With this approach we can share common logic, like error handling, rate limiting, etc. between different datasources, instead of having to reimplement it each time we write a new datasource. Currently there's nothing shared, as it's only 1 datasource, but the interesting stuff will come with the 2nd datasource!

### CODE REVIEW

File: lib/datasource/index.spec.ts [hunk 1]
```diff
@@ -30,7 +30,7 @@ describe('datasource/index', () => {
     expect(datasource.getDatasources()).toBeDefined();
     expect(datasource.getDatasourceList()).toBeDefined();
   });
-  it('validates dataource', async () => {
+  it('validates datsource', async () => {
```
Comment (viceice) [inline lib/datasource/index.spec.ts:33]:
```suggestion
  it('validates datasource', async () => {
```
🙃

File: lib/datasource/cdnjs/index.ts [hunk 2]
```diff
@@ -27,38 +28,42 @@ async function downloadLibrary(library: string): CachePromise<CdnjsResponse> {
...
-      cb: downloadLibrary,
-    });
-    if (!assets) {
-      return null;
-    }
-    const assetName = lookupName.replace(`${library}/`, '');
-    const releases = assets
-      .filter(({ files }) => files.includes(assetName))
-      .map(({ version, sri }) => ({ version, newDigest: sri[assetName] }));
+export class CdnJs extends Datasource {
```
Comment (viceice) [inline lib/datasource/cdnjs/index.ts:31]: I like the class based idea, we should export the class as `default` or with a fixed alias, so we can generate the imports as we now do.

See `versioning` for sample. I think we need to undo the dynamic datasource loading to be like the other modules.

**Review by viceice — COMMENTED** [review]
Rate limiting should be on http layer, not on datasource layer.

**Comment by rarkins** [comment]
Related:
- #5557 - De-duplicate datasource error handling
- #6757 - Add private/auth awareness to datasource index

Also, I'd been thinking the potentially the datasource index could instantiate a `http` instance for each datasource at startup (being sure to include the `id`) and then pass it as part of parameters to `getReleases` or `getDigest`.

File: lib/datasource/cdnjs/index.ts (hunk 2 above)
Comment (JamieMagee) [reply lib/datasource/cdnjs/index.ts:31]: I can export as default no problem.

What was the problem the dynamic generation/loading was trying to solve? I understand the alternative is to maintain the list in `lib/datasource/index.ts` manually, but this list doesn't really change that often.

Comment (viceice) [reply lib/datasource/cdnjs/index.ts:31]: it was to do not have conflicts when users adding new versionigs / datasources / managers ...

Comment (JamieMagee) [reply lib/datasource/cdnjs/index.ts:31]: Oops. I didn't get a notification for your comment.

Is that such a common issue? I feel like it's a rare problem?

Comment (viceice) [reply lib/datasource/cdnjs/index.ts:31]: No, shouldn't be a big problem.

@rarkins can / should we go back to static imports and remove the code generation?

**Comment by viceice** [comment]
? review or not review? 🤔

**Comment by rarkins** [comment]
> ? review or not review? 🤔

Sorry, requested review of wrong branch

File: lib/datasource/cdnjs/index.ts (hunk 2 above)
Comment (rarkins) [reply lib/datasource/cdnjs/index.ts:24]: One remaining problem with maintaining a list was that it wouldn't support dynamic addition of datasources in future.

Comment (viceice) [reply lib/datasource/cdnjs/index.ts:24]: As long as we have the function for returning the full datasource map, we can simply add new datasources.

You can see it in action [here](https://github.com/renovatebot/internal-tools/blob/15413767c5beffaebfe172fe72d1c2063b8ec157/src/utils/datasource/index.ts#L1-L9).
It's used by our internal tools. 🙃

Comment (viceice) [reply lib/datasource/cdnjs/index.ts:24]: If we maintain a static list, I prefer to move it to a separate file, as it will be very long.

Comment (rarkins) [reply lib/datasource/cdnjs/index.ts:24]: I meant for end users to conveniently add custom datasources. If such custom ones need to be pre-configured in an index file, it makes it harder.

Comment (viceice) [reply lib/datasource/cdnjs/index.ts:24]: If a user needs to add custom datasources i would suggest him to use the approach used by our internal-tools.

**Comment by JamieMagee** [comment]
Rebased on top of current `HEAD`. Reverted `crate` changes as this branch has strayed too far from the existing implementation, and rewriting would be simpler than rebasing.

File: lib/datasource/clojure/index.ts [hunk 3]
```diff
@@ -1,8 +1,26 @@
+import { Datasource } from '../datasource';
+import { getReleases } from '../maven';
 import { MAVEN_REPO } from '../maven/common';
+import { GetReleasesConfig, ReleaseResult } from '../types';
```
Comment (viceice) [inline lib/datasource/clojure/index.ts:4]:
```suggestion
import type { GetReleasesConfig, ReleaseResult } from '../types';
```

File: lib/datasource/cdnjs/index.ts [hunk 4]
```diff
@@ -24,38 +22,43 @@ interface CdnjsResponse {
...
-    ).body;
-    if (!assets) {
-      return null;
-    }
-    const assetName = lookupName.replace(`${library}/`, '');
-    const releases = assets
-      .filter(({ files }) => files.includes(assetName))
-      .map(({ version, sri }) => ({ version, newDigest: sri[assetName] }));
+export class CdnJs extends Datasource {
+  readonly id = 'cdnjs';
```
Comment (viceice) [inline lib/datasource/cdnjs/index.ts:26]: can we share this duplicated id somehow?

File: lib/datasource/clojure/index.ts [hunk 5]
```diff
@@ -1,8 +1,26 @@
...
 
-export { getReleases } from '../maven';
+export class Clojure extends Datasource {
+  readonly id = 'clojure';
+
+  readonly registryStrategy = 'merge';
+
+  readonly customRegistrySupport = true;
+
+  readonly defaultRegistryUrls = ['https://clojars.org/repo', MAVEN_REPO];
```
Comment (viceice) [inline lib/datasource/clojure/index.ts:18]: do we need those duplicates?

File: lib/datasource/datasource.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
+import { ExternalHostError } from '../types/errors/external-host-error';
+import { Http, HttpError } from '../util/http';
+import {
```
Comment (viceice) [inline lib/datasource/datasource.ts:3]:
```suggestion
import type {
```

File: lib/datasource/datasource.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+  defaultRegistryUrls: string[];
+
+  defaultVersioning: string;
+
+  registryStrategy: 'first' | 'hunt' | 'merge';
+
+  getId(): string {
+    return this.id;
+  }
```
Comment (viceice) [inline lib/datasource/datasource.ts:31]: Do we need this wrapper function? i think we can directly use `this.id`

File: lib/datasource/index.spec.ts [hunk 8]
```diff
@@ -32,9 +32,8 @@ describe(getName(__filename), () => {
   });
   it('returns datasources', () => {
     expect(datasource.getDatasources()).toBeDefined();
-    expect(datasource.getDatasourceList()).toBeDefined();
   });
-  it('validates dataource', () => {
+  it('validates datsource', () => {
```
Comment (viceice) [inline lib/datasource/index.spec.ts:36]:
```suggestion
  it('validates datasource', () => {
```
🤭

File: lib/datasource/index.spec.ts [hunk 9]
```diff
@@ -32,9 +32,8 @@ describe(getName(__filename), () => {
   });
   it('returns datasources', () => {
     expect(datasource.getDatasources()).toBeDefined();
-    expect(datasource.getDatasourceList()).toBeDefined();
```
Comment (viceice) [inline lib/datasource/index.spec.ts:35]: Why this change?

File: lib/datasource/index.ts [hunk 10]
```diff
@@ -27,7 +29,10 @@ export const getDatasourceList = (): string[] => Array.from(datasources.keys());
 
 const cacheNamespace = 'datasource-releases';
 
-function load(datasource: string): DatasourceApi {
+datasources.set('cdnjs', new CdnJs());
+datasources.set('clojure', new Clojure());
```
Comment (viceice) [inline lib/datasource/index.ts:33]:
```suggestion
```

already done in `./api.ts`. wrong merge?

File: lib/datasource/pod/index.ts [hunk 11]
```diff
@@ -42,7 +42,7 @@ function releasesGithubUrl(
 function handleError(lookupName: string, err: HttpError): void {
   const errorData = { lookupName, err };
 
-  const statusCode = err.response?.statusCode;
+  const statusCode = err.statusCode;
```
Comment (viceice) [inline lib/datasource/pod/index.ts:45]: revert `err.statusCode` is an old `got` property which we like to get rid of.

File: lib/datasource/datasource.ts [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+  protected handleGenericErrors(err: HttpError): never {
+    this.handleSpecificErrors(err);
+    if (err.statusCode !== undefined) {
+      if (
+        err.statusCode === 429 ||
+        (err.statusCode >= 500 && err.statusCode < 600)
+      ) {
+        throw new ExternalHostError(err);
+      }
+    }
```
Comment (viceice) [inline lib/datasource/datasource.ts:52]:
```suggestion
    if (err.response?.statusCode !== undefined) {
      if (
        err.response?.statusCode === 429 ||
        (err.response?.statusCode >= 500 && err.response?.statusCode < 600)
      ) {
        throw new ExternalHostError(err);
      }
    }
```

`statusCode` is a legacy `got` property, should no longer be used.

File: lib/datasource/index.ts [hunk 13]
```diff
@@ -9,6 +9,8 @@ import { clone } from '../util/clone';
 import { regEx } from '../util/regex';
 import * as allVersioning from '../versioning';
 import datasources from './api';
+import { CdnJs } from './cdnjs';
+import { Clojure } from './clojure';
```
Comment (viceice) [inline lib/datasource/index.ts:13]:
```suggestion
```

Not needed, see below

File: lib/util/http/index.ts [hunk 14]
```diff
@@ -47,6 +47,10 @@ export interface HttpResponse<T = string> {
   authorization?: boolean;
 }
 
+export interface HttpError extends Error {
+  statusCode: number;
+}
```
Comment (viceice) [inline lib/util/http/index.ts:52]: Use `HttpError` from `lib/util/types.ts`
```suggestion
```

See other comments for why not use `statusCode` 😉

File: lib/versioning/loose/generic.ts [hunk 15]
```diff
@@ -189,7 +189,6 @@ export abstract class GenericVersioningApi<
     return versions.find((v) => this.equals(v, range)) || null;
   }
 
-  // eslint-disable-next-line class-methods-use-this
```
Comment (viceice) [inline lib/versioning/loose/generic.ts:192]: revert, as you like to revert the eslint setting?

File: package.json [hunk 16]
```diff
@@ -7,8 +7,8 @@
     "renovate-config-validator": "dist/config-validator.js"
   },
   "scripts": {
-    "build": "run-s clean generate:* compile:*",
-    "clean": "rimraf dist",
+    "build": "run-s clean prepare:* generate:* compile:*",
+    "clean": "rimraf dist tmp",
```
Comment (viceice) [inline package.json:11]: Wrong merge?

File: tools/jest-gh-reporter.ts [hunk 17]
```diff
@@ -45,7 +45,6 @@ function getPos(msg: string): Record<string, string> {
 }
 
 class GitHubReporter extends BaseReporter {
-  // eslint-disable-next-line class-methods-use-this
```
Comment (viceice) [inline tools/jest-gh-reporter.ts:48]: wrong?

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/datasource/index.ts (hunk 10 above)
Comment (JamieMagee) [reply lib/datasource/index.ts:33]: Yes, I think this was done before before the dynamic to static migration.

Comment (JamieMagee) [reply lib/datasource/index.ts:33]: Yes, I think this was done before the dynamic to static migration

File: lib/datasource/cdnjs/index.ts [hunk 18]
```diff
@@ -24,38 +22,43 @@ interface CdnjsResponse {
   assets?: CdnjsAsset[];
 }
 
-export async function getReleases({
-  lookupName,
-  registryUrl,
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply lib/datasource/cdnjs/index.ts:29]: Yes. That's the trouble with long-lived branches :cry:

File: lib/datasource/clojure/index.ts (hunk 5 above)
Comment (JamieMagee) [reply lib/datasource/clojure/index.ts:18]: This is an issue I don't yet have a great solution for. If I look at `id`, it's defined in the `DatasourceApi` interface, and I defined it as `abstract id` in the abstract class `Datasource` so that subclasses are required to implement it. However that means that you need to instantiate `ClojureDatasource` in order to read its id.

Maybe dropping the `abstract` modifier, and just making it `static`? [Here's a TS playground link to highlight the issue](https://www.typescriptlang.org/play?#code/IYIwzgLgTsDGEAJYBthjAgIsCaD2ArlLAKYIDeAUAjQpDgJawJQnAAmeAdsgJ4IN2ALjrQGXAOaUAvpUoo0GAMLI8AKyIlsuMIWJkSADwgku7DNvxFSFarXoQmA9ggC8CAOQp1mjwG4ZOQV0BAAxPCgJPAhLXWsDY1NzLBwrfVtA+W5dZBIAOlUJAAoVH1ZYvVI8wQBKLK4c-MKi8Mjoivjq9hqgA)

Comment (JamieMagee) [reply lib/datasource/clojure/index.ts:18]: Or just instantiating the class?

File: lib/datasource/datasource.ts (hunk 7 above)
Comment (JamieMagee) [reply lib/datasource/datasource.ts:31]: It's not possible to use properties directly in the constructor

File: lib/datasource/index.spec.ts (hunk 8 above)
Comment (JamieMagee) [reply lib/datasource/index.spec.ts:36]: Trading 1 spelling mistake for another :smile:

File: .eslintrc.js [hunk 19]
```diff
@@ -42,6 +42,9 @@ module.exports = {
     'prefer-template': 0,
     'no-underscore-dangle': 0,
     'no-negated-condition': 'error',
+    // TODO: re-enable once datasource class migration
+    'class-methods-use-this': 0,
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply .eslintrc.js:46]: Sure:

`getReleases` in `clojure` doesn't use `this` as it uses `getReleases` from `maven` instead of using its own implementation.

`handleSpecificErrors` in the abstract class `Datasource` is a no-op, and not marked as abstract, as not all datasources may need to handle specific errors, so `handleSpecificErrors` is flagged.

I figured it was easier to disable this here during migration, and it can then be re-enabled for specific methods later? I can flag those specific methods now, if you think it's easier or better.

File: lib/datasource/index.spec.ts (hunk 9 above)
Comment (JamieMagee) [reply lib/datasource/index.spec.ts:35]: I think it was a workaround when I had directly imported the datasources in `index.ts`. Reverted.

File: .eslintrc.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (viceice) [reply .eslintrc.js:46]: i think it better to flag the causing code, otherwise you need a lot effort to fix it again

File: lib/datasource/datasource.ts (hunk 7 above)
Comment (viceice) [reply lib/datasource/datasource.ts:31]: Then i think it's better to pass id via constructor and making it readonly in datasource base class

File: lib/manager/deps-edn/extract.ts [hunk 20]
```diff
@@ -16,7 +16,7 @@ export function extractPackageFile(content: string): PackageFile {
     match = regex.exec(rest);
 
     deps.push({
-      datasource: datasourceClojure.id,
+      datasource: new ClojureDatasource().id,
```
Comment (viceice) [inline lib/manager/deps-edn/extract.ts:19]: we should export the id also directly, so we do not need those changes 😕 Looks poretty bad to instanciate for id property only

File: lib/datasource/index.spec.ts [hunk 21]
```diff
@@ -44,9 +44,20 @@ describe(getName(__filename), () => {
       }
       return true;
     }
+    function filterClassBasedDatasources(name: string): boolean {
+      return !['cdnjs', 'clojure'].includes(name);
+    }
```
Comment (viceice) [inline lib/datasource/index.spec.ts:49]: better use `getDatasources()` and check `obj instanceOf Datasource`, so we do not need to change the test on every datasource migration.

**Review by viceice — COMMENTED** [review]
I think we should only convert this two datasources with this pr and convert others in separate pr's, so it's easier to review

File: .eslintrc.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (JamieMagee) [reply .eslintrc.js:46]: Okay, removed this from here, and added it to each individual instance.

File: lib/manager/deps-edn/extract.ts (hunk 20 above)
Comment (JamieMagee) [reply lib/manager/deps-edn/extract.ts:19]: I _think_ I have a better solution for this. I define `Datasource` as

```ts
export abstract class Datasource implements DatasourceApi {
  protected constructor(id: string) {
    this.http = new Http(id);
    this.id = id;
  }
}
```

and implement it like so:

```ts
export class CdnJsDatasource extends Datasource {
  static readonly id = 'cdnjs';

  constructor() {
    super(CdnJsDatasource.id);
  }
}
```

The main issue is this doesn't actually require `id` to be defined. I could put an assert in the constructor that would be caught in unit tests, but throwing exceptions in constructors is generally considered bad practice.

I can't think of a good way to get the compiler to enforce this constraint.

Comment (viceice) [reply lib/manager/deps-edn/extract.ts:19]: Would need strict null check in typescript. But that would need a lot of work. 😉

Comment (viceice) [reply lib/manager/deps-edn/extract.ts:19]: Simpler:

```ts
export abstract class Datasource implements DatasourceApi {
  protected http: Http;
  protected constructor(protected readonly id: string) {
    this.http = new Http(id);
  }
}
```

File: lib/datasource/index.spec.ts (hunk 21 above)
Comment (JamieMagee) [reply lib/datasource/index.spec.ts:49]: Will come back to this later.

I can't directly modify `datasource.getDatasources()` and `clone` is returning `{}` for some reason 🤔

File: .eslintrc.js [hunk 22]
```diff
@@ -42,6 +42,7 @@ module.exports = {
     'prefer-template': 0,
     'no-underscore-dangle': 0,
     'no-negated-condition': 'error',
+
```
Comment (viceice) [inline .eslintrc.js:45]: 😏

File: lib/datasource/datasource.ts [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+  protected http: Http;
+
+  abstract getReleases(
+    getReleasesConfig: GetReleasesConfig
+  ): Promise<ReleaseResult | null>;
+
+  getDigest?(config: DigestConfig, newValue?: string): Promise<string>;
+
+  // eslint-disable-next-line class-methods-use-this
+  handleSpecificErrors(err: HttpError): void {}
```
Comment (viceice) [inline lib/datasource/datasource.ts:37]: We should define an order to we chave consistency. 🤔

File: lib/datasource/datasource.ts [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+
+  defaultConfig: Record<string, unknown>;
+
+  customRegistrySupport: boolean;
+
+  defaultRegistryUrls: string[];
+
+  defaultVersioning: string;
+
+  registryStrategy: 'first' | 'hunt' | 'merge';
```
Comment (viceice) [inline lib/datasource/datasource.ts:26]: maybe add default value?

File: lib/datasource/datasource.ts [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+export abstract class Datasource implements DatasourceApi {
+  protected constructor(public readonly id: string) {
+    this.http = new Http(id);
+  }
+
+  caching: boolean;
+
+  defaultConfig: Record<string, unknown>;
+
+  customRegistrySupport: boolean;
```
Comment (viceice) [inline lib/datasource/datasource.ts:20]: defaults to false?

File: lib/datasource/index.spec.ts [hunk 26]
```diff
@@ -34,19 +34,27 @@ describe(getName(), () => {
...
-      if (module.id !== name) {
-        return false;
-      }
-      return true;
+      return module.id === name;
+    }
+    function filterClassBasedDatasources(name: string): boolean {
+      return !['cdnjs', 'clojure'].includes(name);
     }
     const dss = datasource.getDatasources();
```
Comment (viceice) [inline lib/datasource/index.spec.ts:47]: dss is a `Map`, so clone like this:

```suggestion
    const dss = new Map(datasource.getDatasources());
```

File: lib/datasource/index.spec.ts (hunk 21 above)
Comment (viceice) [reply lib/datasource/index.spec.ts]: clone with `new Map(datasource.getDatasources())` ? 😏

File: lib/datasource/index.spec.ts (hunk 26 above)
Comment (viceice) [reply lib/datasource/index.spec.ts:47]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/Map

File: lib/datasource/datasource.ts (hunk 23 above)
Comment (JamieMagee) [reply lib/datasource/datasource.ts:37]: An order for what?

File: lib/datasource/datasource.ts (hunk 24 above)
Comment (JamieMagee) [reply lib/datasource/datasource.ts:26]: Default should be `first`?

File: lib/datasource/datasource.ts (hunk 25 above)
Comment (JamieMagee) [reply lib/datasource/datasource.ts:20]: Matches the current `DatasourceApi`. That's correct?

Comment (viceice) [reply lib/datasource/datasource.ts:20]: a quick seach shows this is truw for most datasources, so let's make this true by default, so only some datasources need to override :)

File: lib/datasource/datasource.ts (hunk 24 above)
Comment (viceice) [reply lib/datasource/datasource.ts:26]: yes, looks like most registries use `first`, so let's configure as default.

File: lib/datasource/datasource.ts (hunk 23 above)
Comment (viceice) [reply lib/datasource/datasource.ts:37]: member type order and alphabetic naming. maybe we need to open an issue to discuss best practices for us?

**Review by viceice — COMMENTED** [review]
needs lint fix 😉

File: lib/datasource/datasource.ts (hunk 23 above)
Comment (JamieMagee) [reply lib/datasource/datasource.ts:37]: Not sure it's that big a deal right now. Maybe later if this becomes more busy.

For now I've sorted alphabetically, with `protected http` at the bottom as it's a different modifier.

Comment (viceice) [reply lib/datasource/datasource.ts]: you are right, I've opened an issue to track this, so we can make a desision in future. ref #9948

File: lib/datasource/datasource.ts [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+  handleSpecificErrors(err: HttpError): void {}
+
+  protected handleGenericErrors(err: HttpError): never {
+    this.handleSpecificErrors(err);
+    if (err.response?.statusCode !== undefined) {
+      if (
+        err.response?.statusCode === 429 ||
+        (err.response?.statusCode >= 500 && err.response?.statusCode < 600)
+      ) {
+        throw new ExternalHostError(err);
```
Comment (viceice) [inline lib/datasource/datasource.ts:46]: needs coverage 😉

File: lib/datasource/datasource.ts [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+
+  customRegistrySupport = true;
+
+  defaultConfig: Record<string, unknown>;
+
+  defaultRegistryUrls: string[];
+
+  defaultVersioning: string;
+
+  registryStrategy: 'first' | 'hunt' | 'merge' = 'first';
```
Comment (viceice) [inline lib/datasource/datasource.ts:26]:
```suggestion
  caching: boolean | undefined;

  customRegistrySupport = true;

  defaultConfig: Record<string, unknown> | undefined;

  defaultRegistryUrls: string[] | undefined;

  defaultVersioning: string | undefined;

  registryStrategy: 'first' | 'hunt' | 'merge' | undefined = 'first';
```

They should stay optional and we want to make future prove for typescript `strictNullCheck`.

File: lib/datasource/datasource.ts (hunk 27 above)
Comment (JamieMagee) [reply lib/datasource/datasource.ts:46]: It turns out that neither of the 2 datasources I converted actually use this logic. I think either `cargo` must have used it (before I dropped it due to conflicts) or I took it as a decent starting point for generic datasource error handling.

**Comment by JamieMagee** [comment]
@viceice I think I've addressed all your comments.

@rarkins I think this is ready for your review.

**Comment by JamieMagee** [comment]
Pls no more conflicts 😨 🤞

File: lib/datasource/datasource.spec.ts [hunk 29]
```diff
@@ -0,0 +1,45 @@
+import * as httpMock from '../../test/http-mock';
+import { getName } from '../../test/util';
+import { EXTERNAL_HOST_ERROR } from '../constants/error-messages';
+import { Datasource } from './datasource';
+import { GetReleasesConfig, ReleaseResult } from './types';
```
Comment (viceice) [inline lib/datasource/datasource.spec.ts:5]:
```suggestion
import type { GetReleasesConfig, ReleaseResult } from './types';
```

File: lib/datasource/clojure/index.spec.ts [hunk 30]
```diff
@@ -0,0 +1,265 @@
+import _fs from 'fs-extra';
+import upath from 'upath';
+import { ReleaseResult, getPkgReleases } from '..';
+import * as httpMock from '../../../test/http-mock';
+import { getName, loadFixture, mocked } from '../../../test/util';
+import * as hostRules from '../../util/host-rules';
+import { id as versioning } from '../../versioning/maven';
+import { ClojureDatasource as datasource } from '.';
```
Comment (viceice) [inline lib/datasource/clojure/index.spec.ts:8]: Why renaming? Makes code harder to read.

File: lib/datasource/datasource.spec.ts [hunk 31]
```diff
@@ -0,0 +1,45 @@
...
+
+const exampleUrl = 'https://example.com/';
+
+class TestDatasource extends Datasource {
+  constructor() {
+    super('test');
+  }
+
+  // eslint-disable-next-line class-methods-use-this
```
Comment (viceice) [inline lib/datasource/datasource.spec.ts:14]:
```suggestion
```

we are using `this` 😏

**Review by viceice — COMMENTED** [review]
Just some small stuff

**Review by viceice — APPROVED** [review]
🥳🥳
